### ISSUE TITLE
regexp recursion

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8636]
Migrated from [rt.perl.org#40495](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=40495) (status was 'resolved')

Searchable as RT40495$

**Comment by p5pRT** [comment]
### From smurf@smurf.noris.de

#### Created by smurf@smurf.noris.de



$ perl \-le '\("x"x30000\) =~ /\(?&#8203;:0|x\{2\}\)\*/'
Segmentation fault
$

The fault occurs because the the regexp interpreter recurses into
itself\, and runs into the stack resource limit\.

Anchoring the regexp doesn't help\.

Replacing the \{2\} with another copy of x does help\, but if "x" is
replaced with a more complex subexpression things get unwieldy\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.8:

Configured by Debian Project at Fri Jul  7 18:20:22 UTC 2006.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=linux, osvers=2.6.15.7, archname=i486-linux-gnu-thread-multi
    uname='linux rothera 2.6.15.7 #1 smp tue jun 27 18:34:43 utc 2006 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i486-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8 -Darchlib=/usr/lib/perl/5.8 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.8 -Dsitearch=/usr/local/lib/perl/5.8.8 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.8 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.1.2 20060613 (prerelease) (Ubuntu 4.1.1-2ubuntu5)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.4.so, so=so, useshrplib=true, libperl=libperl.so.5.8.8
    gnulibc_version='2.4'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.8:
    /etc/perl
    /usr/local/lib/perl/5.8.8
    /usr/local/share/perl/5.8.8
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8
    /usr/share/perl/5.8
    /usr/local/lib/site_perl
    /usr/local/lib/perl/5.8.4
    /usr/local/share/perl/5.8.4
    .


Environment for perl v5.8.8:
    HOME=/daten/home/smurf
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_CTYPE=de_DE.UTF-8
    LC_MONETARY=de_DE.UTF-8
    LC_TIME=de_DE.UTF-8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=~/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/bin/X11:/usr/games:/usr/X11R6/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @demerphq

On 10/10/06\, via RT Matthias Urlichs \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  Matthias Urlichs
> \# Please include the string&#8203;:  \[perl \#40495\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=40495 >
>
>
>
> This is a bug report for perl from smurf@&#8203;smurf\.noris\.de\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
>
> $ perl \-le '\("x"x30000\) =~ /\(?&#8203;:0|x\{2\}\)\*/'
> Segmentation fault
> $
>
> The fault occurs because the the regexp interpreter recurses into
> itself\, and runs into the stack resource limit\.
>
> Anchoring the regexp doesn't help\.
>
> Replacing the \{2\} with another copy of x does help\, but if "x" is
> replaced with a more complex subexpression things get unwieldy\.

This is fixed in perl 5\.10\, although a pattern like this is going to
consume a lot of memory\, perl doesnt use a DFA so a pattern like this
has to store a lot of information to facilitate backtracking\.

Yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @smpeters

On Tue\, Oct 10\, 2006 at 06&#8203;:37&#8203;:26AM \-0700\, Matthias Urlichs wrote&#8203;:
> \# New Ticket Created by  Matthias Urlichs 
> \# Please include the string&#8203;:  \[perl \#40495\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=40495 >
> 
> 
> 
> This is a bug report for perl from smurf@&#8203;smurf\.noris\.de\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> $ perl \-le '\("x"x30000\) =~ /\(?&#8203;:0|x\{2\}\)\*/'
> Segmentation fault
> $
> 
> The fault occurs because the the regexp interpreter recurses into
> itself\, and runs into the stack resource limit\.
> 
> Anchoring the regexp doesn't help\.
> 
> Replacing the \{2\} with another copy of x does help\, but if "x" is
> replaced with a more complex subexpression things get unwieldy\.
> 

Thanks for the report\.  This problem has been resolved in the development
versions of Perl\.

Steve Peters
steve@&#8203;fisharerojo\.org

**Comment by p5pRT** [comment]
### From smurf@smurf.noris.de

Just for clarification&#8203;: Is the problem fixed by replacing the system
stack recursion in Perl's pattern matcher with an open\-coded one \(which
would still eat lots of memory\)\, or is the problem fixed by simply not
recursing at all?

This pattern is deterministic\. It says \{2\}\, not \{1\,2\} or something like
that\. There's no need to recurse in the first place\.

In a nutshell\, my \(reasonable\, IMHO\) expectation is that x\{N\} should
behave exactly the same way as repeating x N times\. Crashing and/or
allocating O\(len\(string\)\) recursion frames does not do that\.

**Comment by p5pRT** [comment]
smurf@smurf.noris.de - Status changed from 'resolved' to 'open'

**Comment by p5pRT** [comment]
### From smurf@smurf.noris.de

Hi\,

yves orton via RT&#8203;:
> > $ perl \-le '\("x"x30000\) =~ /\(?&#8203;:0|x\{2\}\)\*/'
> > Segmentation fault
> > $
> 
> This is fixed in perl 5\.10\, although a pattern like this is going to
> consume a lot of memory\, perl doesnt use a DFA so a pattern like this
> has to store a lot of information to facilitate backtracking\.
> 
Umm\. Why should Perl want to backtrack anything\, for this pattern?
It's deterministic\.

Hint&#8203;: Replacing the "x\{2\}" with "xx" makes the problem go away\.

\-\- 
Matthias Urlichs   |   \{M&#8203;:U\} IT Design @&#8203; m\-u\-it\.de   |  smurf@&#8203;smurf\.noris\.de
Disclaimer&#8203;: The quote was selected randomly\. Really\. | http&#8203;://smurf\.noris\.de
 \- \-
"The alleged short\-cut to knowledge\, which is faith\,
 is only a short circuit destroying the mind\."
&nbsp;                 \[Ayn Rand\]

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Oct 10\, 2006 at 05&#8203;:02&#8203;:59PM \+0200\, Matthias Urlichs wrote&#8203;:
> Hi\,
> 
> yves orton via RT&#8203;:
> > > $ perl \-le '\("x"x30000\) =~ /\(?&#8203;:0|x\{2\}\)\*/'
> > > Segmentation fault
> > > $
> > 
> > This is fixed in perl 5\.10\, although a pattern like this is going to
> > consume a lot of memory\, perl doesnt use a DFA so a pattern like this
> > has to store a lot of information to facilitate backtracking\.
> > 
> Umm\. Why should Perl want to backtrack anything\, for this pattern?
> It's deterministic\.
> 
> Hint&#8203;: Replacing the "x\{2\}" with "xx" makes the problem go away\.

Hint&#8203;: change the pattern to something like /\(?&#8203;:\[xy\]|x\{2\}\)\*/; in this case
the pattern clearly requires backtracking\. Sometimes the optimiser can
simplify the pattern\, but its not perfect\.

\-\- 
A power surge on the Bridge is rapidly and correctly diagnosed as a faulty
capacitor by the highly\-trained and competent engineering staff\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#9

**Comment by p5pRT** [comment]
### From smurf@smurf.noris.de

On Tue Oct 10 11&#8203;:56&#8203;:34 2006\, davem wrote&#8203;:

> Hint&#8203;: change the pattern to something like /\(?&#8203;:\[xy\]|x\{2\}\)\*/; in this case
> the pattern clearly requires backtracking\. Sometimes the optimiser can
> simplify the pattern\, but its not perfect\.
> 
Well\, if somebody wrote your pattern and let it loose on random input\,
I'd roast them over an open fire\. Slowly\. But that's not the issue here
\-\- anybody who expects the optimizer to translate the above to /\[xy\]\*/
needs to go away until we have hard AI\.

I am not at all concerned with patterns that do need recursion no matter
how you write them\. I \*am\* concerned with patterns that do \*not\* need
it\. Specifically\, I have a problem when a pattern becomes recursive just
because somebody rewrites /XXXXX/ to the equivalent /X\{5\}/\.

Which is exactly what happens here\.

**Comment by p5pRT** [comment]
### From @demerphq

On Tue Oct 10 22&#8203;:47&#8203;:00 2006\, smurf wrote&#8203;:
> On Tue Oct 10 11&#8203;:56&#8203;:34 2006\, davem wrote&#8203;:
> 
> > Hint&#8203;: change the pattern to something like /\(?&#8203;:\[xy\]|x\{2\}\)\*/; in this
case
> > the pattern clearly requires backtracking\. Sometimes the optimiser can
> > simplify the pattern\, but its not perfect\.
> > 
> Well\, if somebody wrote your pattern and let it loose on random input\,
> I'd roast them over an open fire\. Slowly\. But that's not the issue here
> \-\- anybody who expects the optimizer to translate the above to /\[xy\]\*/
> needs to go away until we have hard AI\.
> 
> I am not at all concerned with patterns that do need recursion no matter
> how you write them\. I \*am\* concerned with patterns that do \*not\* need
> it\. Specifically\, I have a problem when a pattern becomes recursive just
> because somebody rewrites /XXXXX/ to the equivalent /X\{5\}/\.
> 
> Which is exactly what happens here\.

No\, its not actually\.

What happens here is that the optimiser is not smart enough to turn

&nbsp;   / P | X\{5\} /x

into

&nbsp;   / P | XXXXX /x

And therefore is unable to TRIE the pattern and determine that P and X
are mutually exclusive\.  

However\, the optimiser is smart enough that it doesn't even use the
regex engine for /X\{5\}/\, let alone cause the regex engine to go recursive\.

You are welcome to help teach the optimiser about transformations like
this\, I know we could use the help\, ive been doing some of it\, such as
/\[x\]/ => /x/ but theres lots more to be done\. But please start a new
thread on perl5porters with any such discussion\, replying to this mail
will reopen the ticket\, which I am now closing\.

Cheers\,
Yves

**Comment by p5pRT** [comment]
@demerphq - Status changed from 'open' to 'resolved'
